core/src/Kokkos_Tuners.hpp

```diff
@@ -0,0 +1,218 @@
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_KOKKOS_TUNERS_HPP
+#define KOKKOS_KOKKOS_TUNERS_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Core_fwd.hpp>
+#include <Kokkos_ExecPolicy.hpp>
+#include <impl/Kokkos_Profiling_Interface.hpp>
+
+#include <array>
+#include <tuple>
+#include <string>
+#include <vector>
+
+namespace Kokkos {
+namespace Tools {
+
+namespace Experimental {
+
+// forward declarations
+SetOrRange make_candidate_set(size_t size, int64_t* data);
+size_t declare_output_type(const std::string&,
+                           Kokkos::Tools::Experimental::VariableInfo);
+void request_output_values(size_t, size_t,
+                           Kokkos::Tools::Experimental::VariableValue*);
+void end_context(size_t context_id);
+class TeamSizeTuner {
+ public:
+  inline void declare_output_type(const std::string& name,
+                                  std::vector<int64_t> candidates) {
+    Kokkos::Tools::Experimental::VariableInfo variable_info;
+
+    variable_info.type =
+        Kokkos::Tools::Experimental::ValueType::kokkos_value_int64;
+    variable_info.category =
+        Kokkos::Tools::Experimental::StatisticalCategory::kokkos_value_ratio;
+    variable_info.valueQuantity =
```

Comment:
I don't love the idea of making this a single integer. It has the potential to fool autotuners.  They really want to know that we are tuning two separate variables with constraints.

Reply:
This is the place where things are complicated. We can't expose "a triangular constraint." What the code you're asking for would look like would be one output type for the vector length, and per vector length an output type for team size (the set of valid team sizes varies per vector length). So the code in `tune` would then look like

```c++
  request_value( /** vector_length */ );
  team_size_id = lookup_team_size_output_type(vector_length)
  request_value( team_size_id );
```

It's not infeasible, but it's nontrivial. I'd prefer group expertise on this as to which implementation we want

Reply:
Resolved (ish) in the new implementation